src/Symfony/Component/Encryption/EncryptionInterface.php

```diff
@@ -0,0 +1,156 @@
+     *
+     * When Alice and Bob wants to communicate securely, they share their public keys with
+     * each other. Alice will encrypt a message with Bob's public key. When Bob
+     * receives the message, he will decrypt it with his private key.
+     *
+     *
+     * <code>
+     *     // Bob:
+     *     $bobKey = $encryption->generateKey();
+     *     $bobPublicOnly = $bobKey->extractPublicKey();
+     *     // Bob sends $bobPublicOnly to Alice
+     *
+     *     // Alice:
+     *     $ciphertext = $encryption->encryptFor('input', $bobPublicOnly);
+     *     // Alice sends $ciphertext to Bob
+     *
+     *     // Bob:
+     *     $message = $encryption->decrypt($ciphertext, $bobKey);
+     * </code>
+     *
+     * @param string       $message      plain text version of the message
+     * @param KeyInterface $recipientKey A key with a public key of the recipient
+     *
+     * @return string the output
+     *
+     * @throws EncryptionException
+     * @throws InvalidKeyException
+     */
+    public function encryptFor(string $message, KeyInterface $recipientKey): string;
+
+    /**
+     * Get an encrypted version of the message that only the recipient can read.
+     * The recipient can also verify who sent the message.
+     *
+     * Asymmetric encryption uses a "key pair" ie a public key and a private key.
+     * It is safe to share your public key, but the private key should always be
+     * kept secret.
+     *
+     * When Alice and Bob wants to communicate securely, they share their public keys with
+     * each other. Alice will encrypt a message with keypair [ alice_private, bob_public ].
+     * When Bob receives the message, he will decrypt it with keypair [ bob_private, alice_public ].
+     *
+     * <code>
+     *     // Alice:
+     *     $aliceKey = $encryption->generateKey();
+     *     $alicePublicOnly = $aliceKey->extractPublicKey();
+     *     // Alice sends $alicePublicOnly to Bob
+     *
+     *     // Bob:
+     *     $bobKey = $encryption->generateKey();
+     *     $bobPublicOnly = $bobKey->extractPublicKey();
+     *     // Bob sends $bobPublicOnly to Alice
+     *
+     *     // Alice:
+     *     $ciphertext = $encryption->encryptForAndSign('input', $bobPublicOnly, $aliceKey);
+     *     // Alice sends $ciphertext to Bob
+     *
+     *     // Bob:
+     *     $keypairForReceiving = $bobKey->createKeypair($alicePublicOnly)
+     *     $message = $encryption->decrypt($ciphertext, $keypairForReceiving);
+     * </code>
+     *
+     * @param string       $message      plain text version of the message
+     * @param KeyInterface $recipientKey A public key of the recipient
+     * @param KeyInterface $senderKey    A private key of the sender
+     *
+     * @return string the output
+     *
+     * @throws EncryptionException
+     * @throws InvalidKeyException
+     */
+    public function encryptForAndSign(string $message, KeyInterface $recipientKey, KeyInterface $senderKey): string;
```

Comment:
If we have a dedicated signer interface, then the sodium implementation requires you to have 2 keys. One for encryption and one for signing. 

That is why I dont really like to separate the interface. I also think it brings complexity that is not really needed. I mean, yes, you can sign a plain text just like JWT is doing. But what is the benefit? Is it to reduce the cost of encryption?

Reply:
If we are going to use Sodium AND if we are going to have one api for encryption and one for signing, then we will need two sets of keys. Eg, "encryption key" and "signing key". 

Im currently on the fence for this with the argument that this would be an easy to use abstraction. If you need to "only sign", you are the "advanced use case" and you will have to use sodium directly.

Reply:
Using the same key pair for both encryption and signature is generally a bad idea, as the requirements for signature keys are often different that encryption ones in terms of e.g. backups, rotation or revokation. 
I suggest to reconsider :)

Reply:
Hm. But now you open a new problem of key management that I conveniently excluded from this PR. 

Should we still leave this to the user or should we help them somehow?

Reply:
Sorry but I think it's more than a key management issue. 
From https://libsodium.gitbook.io/doc/quickstart#how-can-i-sign-and-encrypt-using-the-same-key-pair:

> If you really need to sign and encrypt a message, possibly for signatures to be publicly verified, consider using [signcryption](https://github.com/jedisct1/libsodium-signcryption).
[...]
However, this is not recommended and usually not necessary. Prefer using distinct key pairs instead.